conda/gateways/connection/session.py

```diff
@@ -4,38 +4,72 @@
 from logging import getLogger
 from threading import local
 
-from . import (AuthBase, BaseAdapter, HTTPAdapter, Session, _basic_auth_str,
-               extract_cookies_to_jar, get_auth_from_url, get_netrc_auth, Retry)
+from . import (
+    AuthBase,
+    BaseAdapter,
+    Session,
+    _basic_auth_str,
+    extract_cookies_to_jar,
+    get_auth_from_url,
+    get_netrc_auth,
+    Retry,
+)
 from .adapters.ftp import FTPAdapter
 from .adapters.localfs import LocalFSAdapter
 from .adapters.s3 import S3Adapter
 from ..anaconda_client import read_binstar_tokens
 from ...auxlib.ish import dals
 from ...base.constants import CONDA_HOMEPAGE_URL
 from ...base.context import context
-from ...common.url import (add_username_and_password, get_proxy_username_and_pass,
-                           split_anaconda_token, urlparse)
+from ...common.url import (
+    add_username_and_password,
+    get_proxy_username_and_pass,
+    split_anaconda_token,
+    urlparse,
+)
 from ...exceptions import ProxyError
 
+try:
+    from pip._internal.network.session import HTTPAdapter
```

Comment:
Import from pip like this will be a problem since pip's API isn't supported at all, that's why the module is named `_internal`. We can't rely on it in conda sadly, especially in a part that is security sensitive like this feature.

Reply:
Conda does quite a few "try to import from requests / fall back to pip's vendored _internal version". However the main problem with this PR is that we should re-implement (or vendor?) the pip `_internal` classes used.

The classes we import from pip take care of adding a ssl context argument and propagating it all the way down to the lower-level urllib3 layer.